frontend/src/metabase/dashboard/reducers-typed.ts

```diff
@@ -233,3 +237,77 @@ export const parameterValues = createReducer(
     );
   },
 );
+
+export const loadingDashCards = createReducer(
+  INITIAL_DASHBOARD_STATE.loadingDashCards,
+  builder => {
+    builder
+      .addCase(INITIALIZE, state => {
+        state.loadingStatus = "idle";
+      })
+      .addCase<
+        string,
+        {
+          type: string;
+          payload: {
+            currentTime: number;
+            loadingIds: DashCardId[];
+          };
+        }
+      >(FETCH_DASHBOARD_CARD_DATA, (state, action) => {
```

Comment:
I'd like to see we type the action rather than the reducer because if we do this we have to technically repeat the same type where ever we import the same action.

See [how to type action here.](https://redux-toolkit.js.org/usage/usage-with-typescript#createaction)

For example, here's how the `initialize` action could be typed
```ts
export const initialize = createAction(
  INITIALIZE,
  withPayloadType<
    | {
        clearCache?: boolean;
      }
    | undefined
  >(),
);
```

Reply:
Let me check if it works for `redux-action`'s version of `createAction`. I might need to convert these `createAction` calls to `reduxjs` if I'm getting a ton of type issues

Reply:
IIRC we have a mix of dirrect libs and sometimes actions are not written in RTK but with redux-actions lib and types are different there, so we have to type reducers

Reply:
yep, exactly. I'm trying something really quickly here, but it might blow up into something big and I'm not interested in doing something wild

Reply:
as a reference, I changed the import in `frontend/src/metabase/dashboard/actions/core.ts` to use RTK, and with minimal change, I could pass `yarn type-check`. I haven't tried if there would be other errors in runtime though.

Reply:
Anyway, the recommended "modern" way would be to use `createSlice` which creates both the typed reducer + the typed action in one go: https://redux-toolkit.js.org/usage/usage-with-typescript#createslice but that doesn't play super well with our way of splitting actions and reducers in different files ([which is discouraged by redux](https://redux-toolkit.js.org/usage/usage-with-typescript#createslice)).

But I think using that wouldn't be a "quick and easy" refactor now

Reply:
From a quick look, what I understand is that by using `createSlice` with a `reducers` prop, we won't be able to share the action across multiple reducers, to do that it also provides `extraReducers`.

If most actions are shared (I'm not sure if that's the case), then using `createSlice` with no `reducers` but only `extraReducers` might not make much sense. IDK.

Anyway, `createSlice` could definitely work.

Reply:
Do we often use actions in multiple reducers?
I think it's possible to define the action on the "main" slice that uses it and export it and use it on another.
I believe most of our shared actions are async thunks though, which needs to be created outside of `.recuders` anyway

`createSlice` with only `extraReducers` is basically `createReducer`, I would still default to `createSlice` so that's easy to add a "simple" action in case it's needed

Reply:
I think it's about 70/30 in terms of reusing actions. When we reuse actions we reuse them quite a lot

@npretto maybe at some point you can change one of the reducer/actions to a `createSlice` as an example? I think it would be good to start that at some point just to make everything consistent

Reply:
I'd love to but now I remembered that the last time I mentioned createSlice someone told me that would be a RFC as it would change how we structure files 😢

Reply:
There are 4 files that use `createSlice` in total, the most recent one is 2 months old. I think a quick FE channel message would warrant us the use of `createSlice` not everything needs to have its own RFC I think.

Reply:
Ah no, we already have a few usages actually, so we should be good